crates/polars-io/src/csv/buffer.rs

```diff
@@ -31,29 +35,115 @@ impl PrimitiveParser for Float64Type {
     }
 }
 
-impl PrimitiveParser for UInt32Type {
-    #[inline]
-    fn parse(bytes: &[u8]) -> Option<u32> {
-        lexical::parse(bytes).ok()
-    }
-}
-impl PrimitiveParser for UInt64Type {
-    #[inline]
-    fn parse(bytes: &[u8]) -> Option<u64> {
-        lexical::parse(bytes).ok()
-    }
-}
-impl PrimitiveParser for Int32Type {
-    #[inline]
-    fn parse(bytes: &[u8]) -> Option<i32> {
-        lexical::parse(bytes).ok()
-    }
+macro_rules! impl_integer_parser {
+    ($t:ty, $native_t: ty) => {
+        impl PrimitiveParser for $t
+        where
+            $native_t: Zero
+                + One
+                + AddAssign
+                + MulAssign
+                + SubAssign
+                + CheckedAdd
+                + CheckedSub
+                + CheckedMul
+                + MaxNumDigits,
+        {
+            #[inline]
+            fn parse(text: &[u8]) -> Option<Self::Native> {
+                let mut index;
+                let mut number = <$native_t>::zero();
+
+                let (sign, offset) = text
+                    .first()
+                    .and_then(|&byte| Sign::try_from(byte))
+                    .map(|sign| (sign, 1))
+                    .unwrap_or((Sign::Plus, 0));
+
+                index = offset;
+
+                // Having two dedicated loops for both the negative and the nonnegative case is rather
+                // verbose, yet performed up to 40% better then a more terse single loop with
+                // `number += digit * signum`.
+                let value = match sign {
+                    Sign::Plus => {
+                        let max_safe_digits =
```

Comment:
It is not clear to me if this will have constant time. It should be constant computable if I am not mistaken. If we are creating generics, we could also set it on the type?.

Reply:
> If we are creating generics, we could also set it on the type?

Intuitively, I like this idea. But when it came to implementation, I was a bit scratching my ears and cheeks, and I probably didn't find the trick.

Say we want pre-compute this value: `i32::max_num_digits(nth(10))` and set it on the first const generics:

```rust
fn parse_from_radix_10<const MAX_DIGITS: usize, other_generics>{
      // ........balabalabala
}
```

Due to neither `max_num_digits` nor `nth` being `const fn`, the compiler is unhappy and it also preventing me from implementing `nth` as `const fn`.  I don't have much experience in this area, I would like to hear your advice. 🙇

Reply:
Going on holiday today, but I will see if I can find time to tinker with this. 

I think you are correct about the `const` not working, but the generic part should.

Reply:
Yes, I will refactor it into a generic function, then macro only used for dispatching. 

Most importantly: Have a pleasant holiday! 😄